README.md
# Password-Check CLI

- Website to check the strength of all your passwords easily and instantly!

- This is a password strength/stats checker that is made using [Low Budget Password Strength Estimation](https://github.com/dropbox/zxcvbn) ( Learn more from [here](https://www.semanticscholar.org/paper/zxcvbn%3A-Low-Budget-Password-Strength-Estimation-Wheeler/f7403f27b0517be683836f9c1cb8b0f5a5d82b1a) or [here](https://www.usenix.org/conference/usenixsecurity16/technical-sessions/presentation/wheeler#:~:text=zxcvbn%20is%20an%20alternative%20password,suitable%20for%20mitigating%20online%20attacks.) ). This program will also check your password in a database of some leaked passwords.

# Guide

You can use the `--clean` or `-c` tag to print neatly, without colors


changelog.md
# v3.0

- has both a web ui and a command line interface app

# v2.0

- Checks the password in all of the password lists (passwords should not be hashed) in the `pwdcheck/passwordlists/` directory. Every file eneding with a file extension of `.passwords` or `.txt` will be loaded. Passwords in these files should be divided by lines (one password in one single line)
- `404` and `error` pages were created and are being used in the website
- `404.html` is a modified copy (changed the colors) of [Sad Mac 404 Error Page](https://codepen.io/jkantner/pen/aPLWJm)
- CSS and JS are now in the being reffered from the static folder

# v1.0

- First version
- Checking password with Low Budget Password Strength Estimation and displaying the results
- Web User Interface


start.py
import platform
from os import getcwd, listdir, name

import flask

from pwdcheck.server import app, PWD_LIST_CHECK, all_passwords_count

# Show an ASCII Art
# Taken from `figlet`
print(r"""
            ____  ____   ____ 
            |  _ \/ ___| / ___|
            | |_) \___ \| |    
            |  __/ ___) | |___ 
            |_|   |____/ \____|
        Password Strength Checker
        
""")

# To support both windows and linux+mac
SLASH = "\\"
if name == 'posix':
    SLASH = "/"

    # This information is not available for windows
    pt = platform.uname()
    print(f"""+ System Info -
    * System: {pt.system}
    * Node: {pt.node}
    * Release: {pt.release}
    * Version: {pt.version}
    * Machine: {pt.machine}""")
    print("\n")

# Display the versions
print(f"""+ Other Info -
    * Python3 Version: {platform.python_version()}
    * Flask version: {flask.__version__}""")
print("\n")

# Display the available password lists ( in `pwdcheck/passwordlists` directory`)
# these files should end with a `.txt` or `.passwords` file extention
passwords_dir = f"{getcwd()}{SLASH}pwdcheck{SLASH}passwordlists{SLASH}"
password_lists = listdir(passwords_dir)
if int(len(password_lists)) != 0:  # if there are no files available in the directory
    count = 1
    print("+ Password Lists -", "\n    * Total:", len(password_lists))
    for fileName in password_lists:
        if (fileName.lower().endswith("txt")) or (fileName.lower().endswith("passwords")):
            print(f"    {count}: {fileName}")
        count += 1
    print("\n")

# Display the amount of total loaded passwords from locally available password lists; will print nothing if 0
if PWD_LIST_CHECK == True:
    print("+ Loaded all the passwords -",
          "\n    * Count:", all_passwords_count)
    print("\n")

app.run('0.0.0.0', port=6969)


pwdchk.py
import sys

from colorama import init, Fore, Back, Style
from zxcvbn import zxcvbn


def passwordStat(password):
    try:
        results = zxcvbn(f"{password}")
    except Exception as e:
        print("Error:", e)

    sequence_info = ""
    for dic in results["sequence"]:
        for k, v in dic.items():
            if str(k) == "base_matches":
                if isinstance(v[0], dict):
                    sequence_info += "\n- base_matches -\n"
                    for k2, v2 in v[0].items():
                        sequence_info += f"{k2} - {v2}\n"
                    sequence_info += "\n"
            else:
                sequence_info += f"{k} - {v}\n"
        sequence_info += "\n"

    try:
        calc_time = results['calc_time']
    except:
        calc_time = "-"
    try:
        crack_time = f"Online throttling 100 per hour - {results['crack_times_display']['online_throttling_100_per_hour']}\nOnline throttling 10 per second - {results['crack_times_display']['online_no_throttling_10_per_second']}\nOffline slow hasing 1e4 per second - {results['crack_times_display']['offline_slow_hashing_1e4_per_second']}\nOffline fast hasing 1e10 per second - {results['crack_times_display']['offline_fast_hashing_1e10_per_second']}"
    except:
        crack_time = "-"

    warning = "-"
    try:
        if results['feedback']['warning'] == '':
            warning = "-"
        else:
            warning = results['feedback']['warning']
    except:
        warning = "-"

    suggestions = "-"
    try:
        if len(results["feedback"]["suggestions"]) == 0:
            pass
        else:
            suggestions = ""
            for item in results["feedback"]["suggestions"]:
                suggestions += f"{item}\n"
    except:
        suggestions = "-"

    return {
        'details': results,
        'sequence_info': sequence_info,
        'calc_time': calc_time,
        'crack_time': crack_time,
        'warning': warning,
        'suggestions': suggestions
    }


def print_box(text: str):
    lines = text.split('\n')
    max_length = max(len(line) for line in lines)
    border = '+' + '-' * (max_length + 2) + '+'
    print(border)
    for line in lines:
        print('| ' + line.ljust(max_length) + ' |')
    print(border)


def passwordCheck(passwords, clean):
    if len(passwords) == 0:
        tmp = input("Password> ")
        passwords.append(tmp)

    for count, password in enumerate(passwords, 1):
        stat = passwordStat(password=password)
        if clean:
            print_box(
                f"#{count}: {password}\n\n" +
                f"{'#'*29} Score {'#'*29}\n{stat['details']['score']}\n\n\n" +
                f"{'#'*28} Guesses {'#'*28}\n{stat['details']['guesses']}\n\n\n" +
                f"{'#'*27} Crack Time {'#'*26}\n{stat['crack_time']}\n\n\n" +
                f"{'#'*27} Warnings {'#'*28}\n{stat['warning']}\n\n\n" +
                f"{'#'*26} Suggestions {'#'*26}\n{stat['suggestions']}\n\n\n" +
                f"{'#'*25} Sequence Info {'#'*25}\n{stat['sequence_info']}\n\n\n" +
                f"{'#'*66}\n{stat['calc_time']}\n\n\n"
            )
        else:
            _start = f"{Style.RESET_ALL}{Back.MAGENTA}{Fore.MAGENTA}"
            _title_end = f"{Back.MAGENTA}{Fore.MAGENTA}"
            print(f"{Style.RESET_ALL}{Back.YELLOW}{Fore.YELLOW}{'#'*66}{Style.RESET_ALL}")
            print(
                f"### {password} ###\n\n" +
                f"{_start}{'#'*29}{Fore.RESET} Score {_title_end}{'#'*29}{Style.RESET_ALL}\n{stat['details']['score']}\n\n\n" +
                f"{_start}{'#'*28}{Fore.RESET} Guesses {_title_end}{'#'*28}{Style.RESET_ALL}\n{stat['details']['guesses']}\n\n\n" +
                f"{_start}{'#'*27}{Fore.RESET} Crack Time {_title_end}{'#'*26}{Style.RESET_ALL}\n{stat['crack_time']}\n\n\n" +
                f"{_start}{'#'*27}{Fore.RESET} Warnings {_title_end}{'#'*28}{Style.RESET_ALL}\n{stat['warning']}\n\n\n" +
                f"{_start}{'#'*26}{Fore.RESET} Suggestions {_title_end}{'#'*26}{Style.RESET_ALL}\n{stat['suggestions']}\n\n" +
                f"{_start}{'#'*25}{Fore.RESET} Sequence Info {_title_end}{'#'*25}{Style.RESET_ALL}\n{stat['sequence_info']}\n" +
                f"{_start}{'#'*66}{Style.RESET_ALL}\nIt took {stat['calc_time']} seconds\n"
            )


def main():
    args = sys.argv[1:]

    clean = False
    if ("--clean" in args) or ("-c" in args):
        try:
            args.remove("--clean")
        except:
            args.remove("-c")
        clean = True

    passwordCheck(passwords=args, clean=clean)


if __name__ == "__main__":
    init()
    main()


pwdcheck/server.py
import os

from flask import Flask, render_template, request, url_for
from zxcvbn import zxcvbn

# Multi Platform Support
SLASH = "\\"
if os.name == 'posix':
    SLASH = "/"


app = Flask(__name__)


# loading all the passwords from the password lists
# from the files (ending with `.txt` or `.passwords`) in pwdcheck/passwordlists
passwords_dir = f"{os.getcwd()}{SLASH}pwdcheck{SLASH}passwordlists{SLASH}"
password_lists = os.listdir(passwords_dir)
all_passwords = []
for fileName in password_lists:
    try:
        with open(f"{passwords_dir}{fileName}", "r", encoding="utf-8") as f1:
            allLines = f1.readlines()
            for oneLine in allLines:
                cleaned = oneLine.strip()
                if cleaned not in all_passwords:
                    all_passwords.append(cleaned)
            allLines = []
    except:
        pass

# Incase if any password-list does not exist
try:
    all_passwords_count = len(all_passwords)
except:
    all_passwords_count = 0

try:
    if all_passwords_count == 0:
        PWD_LIST_CHECK = False
    else:
        PWD_LIST_CHECK = True
except:
    PWD_LIST_CHECK = False
    all_passwords_count = 0


@app.route('/', methods=['GET', 'POST'])
def index():
    """
    The main web-page

    if the request method is GET, it will return the `index.html` without any data regarding the passwords

    if the request method is POST,
        the password will be passed to the Low Budget Passoword Strength Estimation function (https://pypi.org/project/zxcvbn/)
        if the above step fails, it will return the `error.html` with the exact Python3 Exception and it will be displayed in the webpage
        make everything organized, the "sequence" key from the output is handled to display the `key - value`
        if PWD_LIST_CHECK is True, the password will be cheked in the database(list) and if it exists, it will be displayed
        a POST request will be send to `index.html` after all the data is ready

    Returns:
        - flask.render_template: `error.html` with an error if any
        - flask.render_template: `index.html` with the data
    """
    if request.method == 'GET':
        return render_template("index.html")

    if request.method == 'POST':
        passwordi = request.form.get('password')

        try:
            results = zxcvbn(f"{passwordi}")
        except Exception as pwderr:
            return render_template("error.html",
                                   pwderr=pwderr)

        sequence_info = ""
        for dic in results["sequence"]:
            for k, v in dic.items():
                if str(k) == "base_matches":
                    if isinstance(v[0], dict):
                        sequence_info += "\n- base_matches -\n"
                        for k2, v2 in v[0].items():
                            sequence_info += f"{k2} - {v2}\n"
                        sequence_info += "\n"
                else:
                    sequence_info += f"{k} - {v}\n"
            sequence_info += "\n"

        try:
            calc_time = results['calc_time']
        except:
            calc_time = "-"

        try:
            crack_time = f"Online throttling 100 per hour - {results['crack_times_display']['online_throttling_100_per_hour']}\nOnline throttling 10 per second - {results['crack_times_display']['online_no_throttling_10_per_second']}\nOffline slow hasing 1e4 per second - {results['crack_times_display']['offline_slow_hashing_1e4_per_second']}\nOffline fast hasing 1e10 per second - {results['crack_times_display']['offline_fast_hashing_1e10_per_second']}"
        except:
            crack_time = "-"

        warning = "-"
        try:
            if results['feedback']['warning'] == '':
                pass
            else:
                warning = results['feedback']['warning']
        except:
            warning = "-"

        suggestions = "-"
        try:
            if len(results["feedback"]["suggestions"]) == 0:
                pass
            else:
                suggestions = ""
                for item in results["feedback"]["suggestions"]:
                    suggestions += f"{item}\n"
        except:
            suggestions = "-"

        leaked = "-"
        try:
            if PWD_LIST_CHECK:
                if str(passwordi) in all_passwords:
                    leaked = f"Your password has been found in a database of {all_passwords_count} leaked passwords"
                else:
                    leaked = "-"
        except:
            leaked = "-"

        return render_template(
            'index.html',
            password=results['password'],
            score=results['score'],
            guesses=results['guesses'],
            sequence_info=sequence_info,
            calc_time=calc_time,
            crack_time=crack_time,
            warning=warning,
            suggestions=suggestions,
            leaked=leaked
        )


@app.errorhandler(404)
def page_not_found(e):
    return render_template('404.html'), 404


if __name__ == "__main__":
    app.run('0.0.0.0', port=6969, debug=True)
